pkg/envtest/crd.go

```diff
@@ -0,0 +1,200 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+)
+
+// CRDOptions are the options for installing CRDs
+type CRDOptions struct {
+	// Paths is the path to the directory containing CRDs
+	Paths []string
+
+	// CRDs is a list of CRDs to install
+	CRDs []*apiextensionsv1beta1.CustomResourceDefinition
+
+	// maxTime is the max time to wait
+	maxTime time.Duration
+
+	// pollInterval is the interval to check
+	pollInterval time.Duration
+}
+
+const defaultPollInterval = 100 * time.Millisecond
+const defaultMaxWait = 10 * time.Second
+
+// InstallCRDs installs a collection of CRDs into a cluster by reading the crd yaml files from a directory
+func InstallCRDs(config *rest.Config, options CRDOptions) ([]*apiextensionsv1beta1.CustomResourceDefinition, error) {
+	defaultCRDOptions(&options)
+
+	// Read the CRD yamls into options.CRDs
+	if err := readCRDFiles(&options); err != nil {
+		return options.CRDs, err
+	}
+
+	// Create the CRDs in the apiserver
+	if err := CreateCRDs(config, options.CRDs); err != nil {
+		return options.CRDs, err
+	}
+
+	// Wait for the CRDs to appear as Resources in the apiserver
+	if err := WaitForCRDs(config, options.CRDs, options); err != nil {
+		return options.CRDs, err
+	}
+
+	return options.CRDs, nil
+}
+
+// readCRDFiles reads the directories of CRDs in options.Paths and adds the CRD structs to options.CRDs
+func readCRDFiles(options *CRDOptions) error {
+	if len(options.Paths) > 0 {
+		for _, path := range options.Paths {
+			new, err := readCRDs(path)
+			if err != nil {
+				return err
+			}
+			options.CRDs = append(options.CRDs, new...)
+		}
+	}
+	return nil
+}
+
+// defaultCRDOptions sets the default values for CRDs
+func defaultCRDOptions(o *CRDOptions) {
+	if o.maxTime == 0 {
+		o.maxTime = defaultMaxWait
+	}
+	if o.pollInterval == 0 {
+		o.pollInterval = defaultPollInterval
+	}
+}
+
+// WaitForCRDs waits for the CRDs to appear in discovery
+func WaitForCRDs(config *rest.Config, crds []*apiextensionsv1beta1.CustomResourceDefinition, options CRDOptions) error {
+	// Add each CRD to a map of GroupVersion to Resource
+	waitingFor := map[schema.GroupVersion]*sets.String{}
+	for _, crd := range crds {
+		gv := schema.GroupVersion{Group: crd.Spec.Group, Version: crd.Spec.Version}
+		if _, found := waitingFor[gv]; !found {
+			// Initialize the set
+			waitingFor[gv] = &sets.String{}
+		}
+		// Add the Resource
+		waitingFor[gv].Insert(crd.Spec.Names.Plural)
+	}
+	// Poll until all resources are found
+	return wait.PollImmediate(options.pollInterval, options.maxTime, func() (bool, error) {
+		// Create a new clientset to avoid any client caching of discovery
+		cs, err := clientset.NewForConfig(config)
+		if err != nil {
+			return false, err
+		}
+
+		allFound := true
+		for gv, resources := range waitingFor {
+			// All resources found, do nothing
+			if resources.Len() == 0 {
+				delete(waitingFor, gv)
+				continue
```

Comment:
I wonder if it is safe to delete an element from map while iterating over it ?

Reply:
Yes it is.  I had the same question.

https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-golang-map-within-a-range-loop